sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -2228,3 +2228,13 @@ object AsOfJoin {
     }
   }
 }
+
+/**
+ * Trait that logical plans can extend to check whether it can allow non-deterministic
+ * expressions and pass the CheckAnalysis rule.
+ */
+trait AllowsNonDeterministicExpression extends LogicalPlan {
```

Comment:
This reminds me of `SupportsSubquery`. Shall we follow the same naming convention? `SupportsNonDeterministicExpression`. And we can place this one next to `SupportsSubquery`.